[IMP] website: add qweb view to see view hierarchy

rdeodoo opened the pull request:
The core website behavior is the COW (copy on write) which makes it possible to
handle multiple websites on a single database.

When you do a CRUD operation on an `ir.ui.view` in a website context, the ORM
normal behavior is bypass to perform special stuff, such as:
  - Write: if the view is generic (no `website_id`), we bifurcate the write on
           the specific view (same `key` but with `website_id` set). If that
           view doesn't exist, we create it and fork its whole descending
           hierarchy. This write behavior is known as COW.
  - Create: we ensure that view is created only for the website and no other.
            Basically we just add a `website_id` to that view, but it might
            also be about setting the correct `inherit_id` by finding the
            specific view among multiple views with the same `key`.
  - Unlink: if the view is generic, we will unlink that view but also create a
            copy of that view for every other websites so they remain
            unmodified: this is known as COU (copy on unlink).

All those behavior aims to satisfy to the multi website holy grail:
 -- Any modification in a website context should never impact another website --

Because of those uncommon behaviors, it is sometimes hard to understand or
investigate a database, as views might be copied, up to the point where there
is multiple whole view hierarchies.

This commit attempts to create an user interface to easily figure what's going
on in a database at an `ir.ui.view` level.
The mains advantages of that new `ir.ui.view` Qweb view is to:
  - easily figure which views are COW'd on which website
  - easily detect which views have a real `arch` modification
  - easily see that `arch` change

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/63040).

rdeodoo commented on addons/website/static/src/js/content/website_root.js:
```diff
@@ -254,6 +254,23 @@ var WebsiteRoot = publicRootData.PublicRoot.extend(KeyboardNavigationMixin, {
+     * @private
+     * @param {OdooEvent} ev
+     */
+    _onBtnSpinRequest: function (ev) {
```
Extracted in 14.0 to https://github.com/odoo/odoo/pull/63417
Wait & see if we can remove it from this PR

rdeodoo commented on addons/website/views/website_views.xml:
```diff
@@ -298,6 +303,95 @@
+                    <i class="fa fa-eye ml-2" title="Go to View"/>
+                </a>
+                <!-- TODO button to show diff -->
+                <a type="action" name="base.reset_view_arch_wizard_action"
```
Not too sure yet how to pass active_model and active_ids to the python action, despite the context attribute, it is empty on the python action

rdeodoo commented:
@JKE-be I think everything we discussed has been processed.

The UI is much better now with those small details. The important things (forked/arch updated) are highlighted and you directly see it.

**Done:**
```
- remove search panel
- remove 'Only Generic Views'
- hide inactive views by default
- inverse icons, + when collapsed, - when not collapsed
- remove fold all/unfold all
- click on fold should fold self and all children, click on unfold should only unfold self
- always show tree hierarchy from top level
- pencil icon to show arch is updated is not enough visible
```
`- name (xml_id) [website] as display name`
For this one, I first though to override `name_get()` but it is not convenient if we want to apply style to different part of it (eg key in bold, website in color). But it might still be useful for other flows so I added it.
![image](https://user-images.githubusercontent.com/30048408/102804521-2776b080-43ba-11eb-9968-a754c807a5f7.png)

As we now load tree from top level view, as discussed, I added an highlight and scroll system to show the requested view.
![image](https://user-images.githubusercontent.com/30048408/102804604-41b08e80-43ba-11eb-8f5a-4940f8492071.png)

**Not sure:**
`keyboard nav up/down to navigate and left/right to fold/unfold`
as I am not too sure how it would feels/works with Odoo and/or the browser itself. Right now the `Tab` and `Enter` are just listened inside the search input.
`website 1 by default in the dropdown`
as I feel it would be weird if you wan't to see another website or all websites

**Left to do:** 
- the show diff button (shown on hover) is not working yet
- when clicking on "Multiple tree exists", it redirects you to the form view but it should redirect to the tree view hierarchy instead

Both are related as I didn't figured yet how to load action with active_id set from the qweb view

rdeodoo commented:
- I removed the commit about adding a loading button helper as it will be merged in 14.0 instead, see https://github.com/odoo/odoo/pull/63417
- I applied BSO review:
  - Add `Shift` listener to backward search
  - Show website name and not website id in website dropdown selector
- I applied BVR review:
  - z-index issue with dropdown menu of the systray
  - clear search results when the search input is emptied

I didn't do:
- [BVR] When navigating in views through the tree hierarchy, going back through the breadcrump doesn't keep the website filter or the show inactive filter.
  - -> I think it is high effort low reward to do that for this very advanced view
- [BSO] Complete keyboard navigation
  - -> I think it is high effort low reward to do that for this very advanced view
  
I will focus on making the show diff works and this one will be good to be reviewed

rdeodoo commented:
Last 2 TODOs (show diff & load copied trees) are done :partying_face: 
I rebased and wrote the commit messages, this is finally ready for a review.

rdeodoo commented:
- ![image](https://user-images.githubusercontent.com/30048408/111186208-169d9b80-85b3-11eb-8cff-5c700affd84d.png) dashed angle line is now correctly set if the last view in inactive and hidden
- `Show View Hierarchy` is no more a stat button but a link in the form view next to the `type` field (only if qweb) 
![image](https://user-images.githubusercontent.com/30048408/111202159-dabf0200-85c3-11eb-9f59-508059d0567f.png)
- First website is now selected by default (website filter)

I think we are good here, I don't remember anything else from the review, in case you remember don't hesitate to tell me.
For the view hierarchy we thought about changing the display to regroup cow views but I can't remember how..

rdeodoo commented:
@JKE-be It is possible I forgot some stuff we discussed, maybe it will come back to you if you gives it a quick last test.
We can discuss the merge of cow view/generic view in the tree but I can't remember how we would do that

JKE-be commented:
@robodoo r+ rebase-ff

waiting the next step, delete a specific hierarchy

robodoo commented:
Merge method set to rebase and fast-forward
